dpctl/tensor/_dlpack.pyx

```diff
@@ -0,0 +1,399 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+# distutils: language = c++
+# cython: language_level=3
+# cython: linetrace=True
+
+cimport cpython
+from libc cimport stdlib
+from libc.stdint cimport int32_t, int64_t, uint8_t, uint16_t, uint64_t
+
+cimport dpctl as c_dpctl
+cimport dpctl.memory as c_dpmem
+
+from .._backend cimport (
+    DPCTLDevice_Delete,
+    DPCTLDevice_GetParentDevice,
+    DPCTLSyclDeviceRef,
+    DPCTLSyclUSMRef,
+)
+from ._usmarray cimport usm_ndarray
+
+import numpy as np
+
+import dpctl
+import dpctl.memory as dpmem
+
+
+cdef extern from './include/dlpack/dlpack.h' nogil:
+    cdef int DLPACK_VERSION
+
+    cdef enum DLDeviceType:
+        kDLCPU
+        kDLCUDA
+        kDLCUDAHost
+        kDLCUDAManaged
+        kDLROCM
+        kDLROCMHost
+        kDLOpenCL
+        kDLVulkan
+        kDLMetal
+        kDLVPI
+        kDLOneAPI
+
+    ctypedef struct DLDevice:
+        DLDeviceType device_type
+        int device_id
+
+    cdef enum DLDataTypeCode:
+        kDLInt
+        kDLUInt
+        kDLFloat
+        kDLBfloat
+        kDLComplex
+
+    ctypedef struct DLDataType:
+        uint8_t code
+        uint8_t bits
+        uint16_t lanes
+
+    ctypedef struct DLTensor:
+        void* data
+        DLDevice device
+        int ndim
+        DLDataType dtype
+        int64_t* shape
+        int64_t* strides
+        uint64_t byte_offset
+
+    ctypedef struct DLManagedTensor:
+        DLTensor dl_tensor
+        void* manager_ctx
+        void (*deleter)(DLManagedTensor*)  # noqa: E211
+
+
+def get_build_dlpack_version():
+    """
+    Returns the string value of DLPACK_VERSION from dlpack.h
+    `dpctl.tensor` was built with.
+    """
+    return str(DLPACK_VERSION)
+
+
+cdef void _pycapsule_deleter(object dlt_capsule):
+    cdef DLManagedTensor *dlm_tensor = NULL
+    if cpython.PyCapsule_IsValid(dlt_capsule, 'dltensor'):
+        dlm_tensor = <DLManagedTensor*>cpython.PyCapsule_GetPointer(
+            dlt_capsule, 'dltensor')
+        dlm_tensor.deleter(dlm_tensor)
+
+
+cdef void _managed_tensor_deleter(DLManagedTensor *dlm_tensor) with gil:
+    if dlm_tensor is not NULL:
+        stdlib.free(dlm_tensor.dl_tensor.shape)
+        cpython.Py_DECREF(<usm_ndarray>dlm_tensor.manager_ctx)
+        dlm_tensor.manager_ctx = NULL
+        stdlib.free(dlm_tensor)
+
+
+cpdef to_dlpack_capsule(usm_ndarray usm_ary) except+:
+    """
+    Constructs named Python capsule object referencing
+    instance of `DLManagerTensor` from
+    :class:`dpctl.tensor.usm_ndarray` instance.
+    """
+    cdef c_dpctl.SyclQueue ary_sycl_queue
+    cdef c_dpctl.SyclDevice ary_sycl_device
+    cdef DPCTLSyclDeviceRef pDRef = NULL
+    cdef DLManagedTensor *dlm_tensor = NULL
+    cdef DLTensor* dl_tensor = NULL
+    cdef int nd = usm_ary.get_ndim()
+    cdef char* data_ptr = usm_ary.get_data()
+    cdef Py_ssize_t *shape_ptr = NULL
```

Comment:
Inconsistent styling. Please check our  CAPI clang-format styling for where the `*` is to be placed: next to the variable name or after type.

Reply:
Ok, but this inconsistency is pervasive throughout Cython codebase. Perhaps we should address that in a separate PR.

In `_sycl_queue.pyx` we use `void *x`. I will make the change for `_dlpack.pyx` to do the same.

Reply:
Ok, pushed the change.